docs/RFCS/00000000_cascade.md

```diff
@@ -0,0 +1,815 @@
++------+
+(1 row)
+
+SELECT * FROM c;
++------+
+| a_id |
++------+
+|    1 |
++------+
+(1 row)
+
+SELECT * FROM d;
++--------+--------+
+| b_a_id | c_a_id |
++--------+--------+
+|      1 |      1 |
++--------+--------+
+(1 row)
+
+DELETE FROM a WHERE id = 1;
+
+SELECT * FROM a;
++----+
+| id |
++----+
+|  2 |
++----+
+(1 row)
+
+SELECT * FROM b;
++------+
+| a_id |
++------+
++------+
+(0 rows)
+
+SELECT * FROM c;
++------+
+| a_id |
++------+
+|    2 |
++------+
+(1 row)
+
+SELECT * FROM d;
++--------+--------+
+| b_a_id | c_a_id |
++--------+--------+
++--------+--------+
+(0 rows)
+```
+
+#### Self-Referential Cascading Deletes
+
+Without going into the circular referencing which would required deferring, it is still possible to get self-referential tables that can cascade deletes between rows.
+
+```SQL
+CREATE TABLE a (
+  id INT PRIMARY KEY
+ ,other_id INT REFERENCES a ON DELETE CASCADE
+);
+INSERT INTO a VALUES (1, NULL), (2, 1), (3, 1), (4, 1);
+
+SELECT * FROM a;
++----+----------+
+| id | other_id |
++----+----------+
+|  1 | NULL     |
+|  2 |        1 |
+|  3 |        1 |
+|  4 |        1 |
++----+----------+
+(4 rows)
+
+DELETE FROM a WHERE id = 1;
+
+SELECT * FROM a;
++----+----------+
+| id | other_id |
++----+----------+
++----+----------+
+(0 rows)
+```
+
+### Changes required
+
+*This section is still a work in progress and prototyping some solutions will bear out many more details.*
+
+Almost all of the changes will be in `fk.go` and will need to take into consideration the possibility that we will be adding in deferrable constraints in the future. This does break the general pragmatic rule of *code for what you need today not for what you will need tomorrow*. So the suggestion here is not to alter the design, but to be cognizant of the implications of deferrable.
+
+Most of the implementation details will be decided during the execution phase, but there are two main options.
+
+The first option is to just walk the tree and execute each cascading action immediately, following up on any cascade as it occurs. This should be relatively easy to implement, but may take a long time to fail when there is a constraint violation. It also may also double write on conflicts, but those are an exceptionally rare occurrences.
+
+The second is to pre-determine all cascading actions into one collection while walking the tree, than execute them. This adds the advantage of being able to deal with conflicts quickly and to fail without having to perform the actual writes. Also, it may allow the batching of multiple writes when there are multiple writes to the same table.
+
+#### Interleaved Optimizations
+
+Since one of our primary use cases will be interleaved tables, special attention should be paid to optimizing cascading deletes and updates.
+
+*More details on these optimizations will be added to this document soon.*
+
+### Plan of Engineering
+
+The plan will be to implement the new actions in order of importance, which is roughly:
```

Comment:
For your ordering, you should consider ease of completing the task. For example, it's likely that tackling `ON DELETE NULL` first will make `ON DELETE CASCADE` more tractable. Also, the former is a known issue with Sequelize compatibility, so it has inherent value.